### ISSUE TITLE
Consider Flagging Usage of `!important` in Stylelint

### ISSUE THREAD

**Issue by mikeselander** [humanmade/coding-standards#144]
While `!important` _can_ be used intelligently and useful in specific situations, it's generally bad practice. I know that we don't want to specifically block all usages of `!important`, but I think it would be beneficial to make a developer justify each usage of it.

I recommend that we adapt the [existing `!important` rule](https://github.com/stylelint/stylelint/blob/master/lib/rules/declaration-no-important/index.js) (or just use it as-is) to display a message requiring a justifying comment to dismiss the error. In this way, a reviewer can review the logic for using `!important` which should allow them to offer alternatives and hopefully reduce the usage of `!important`.

Thoughts @humanmade/frontend?

**Comment by rmccue** [comment]
Generally, I've avoided the practice of warnings in favour of either erroring or nothing. The problem with warnings is generally they need to be dismissable, and we don't have a framework for that, so it would cause the warning to be highlighted every time. This adds to the cognitive overhead of reviewing, and generally means people end up trusting the standards less.

With that said, we could potentially have a rule that requires a justification for every `!important` in a comment or something?

**Comment by peterwilsoncc** [comment]
I’d rather not, it’s legitimate in a quite a few cases in the type of css our coding standards encourage people to write. 

Documenting and encouraging only these use cases is of more use than requiring justification of each use.

**Comment by joemcgill** [comment]
Curious if we could write a sniff that looks for an inline comment with any instance of `!important`?

I agree that it should be highly discouraged but not blocked entirely.

**Comment by rmccue** [comment]
Adding the rule and having it be ignored explicitly would work; see [stylelint's own example](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#turning-rules-off-from-within-your-css):

```
#id {
  color: pink !important; /* stylelint-disable-line declaration-no-important */
}
```

I'm unsure if there's a way to require a human-readable comment though, which IMO is what we want.

**Comment by kirstyburgoine** [comment]
Maybe it's worth introducing a `shame` file as a standard to each project and then the rule could instead flag if the `!important` has been used outside of that file? It doesn't solve the exact need to encourage documenting the reason, but because it has had to be added to the shame file, it does force people to think more about whether they really need it and that thought process would (hopefully) be documented in a comment.

**Comment by mattheu** [comment]
There are occasionally times when `!important` has its place, and a hard error does seem a bit severe. 

One example - I wrote custom styles for something that came from a 3rd party script. They had their own styles, that used `!important`. To override these, I had to use `!important` too. Of course the 3rd party should not have used important, but I wouldn't want a developer to spend much time trying to solve this problem without just using important themselves. 

So if the language used by the error was more like, 'check valid use of !important` rather than `never use !important` - that may be better.

I'm also not a fan of just throwing random things into a `shame` file - I'd encourage you to still keep your component code together on one file even if you have to use `!important` for some reason.

>  it’s legitimate in a quite a few cases in the type of css our coding standards encourage people to write.

@peterwilsoncc I'm curious what you mean by this. Could we improve the coding standards?

**Comment by mikeselander** [comment]
Thanks for all the comments, I was hoping this would inspire some discussion ❤️ 

Just to be clear, I am not proposing that we completely disallow using `!important`. I am proposing that we make using it just enough of a pain that developers want to re-consider using it. If it's actually required, the developer leaves a comment outlining why, and if it's not required, the developer will come up with new code.

> With that said, we could potentially have a rule that requires a justification for every !important in a comment or something?

I should have been clearer - this is exactly what I'd like to see. The rule would only flag instances where there is not an ignore or justification comment.

> Documenting and encouraging only these use cases is of more use than requiring justification of each use.

While documenting this is important, I disagree that it's more useful. Developers will read our guidelines once a month or quarter if we're lucky, but they interface with the coding standards on a daily basis. 

> I'm also not a fan of just throwing random things into a shame file - I'd encourage you to still keep your component code together on one file even if you have to use !important for some reason.

I agree with this, styles should generally be as close to their component as possible.

> @peterwilsoncc I'm curious what you mean by this. Could we improve the coding standards?

huge +1 - if we are accidentally encouraging `!important`, we should fix that.

**Comment by kirstyburgoine** [comment]
I'm gonna go out on a limb and disagree with you both about using a shame file.

By needing to add hacky code (i.e. anywhere you use an `!important`), to a shame file it forces you to really think about whether that `!important` is needed. When you add it, it makes it super transparent to anyone who looks at that repo that you know this is hacky, you admit to having to resort to this, but there is a definite reason for its need (which should be commented in the shame file as well).

Keeping hacky code in one place helps isolate code that should be refactored, which is definitely a help when always aiming to leave a code base a little better than when you found it. It also makes it easier for anyone new to the project, later after all of the original developers have moved on to other projects, to identify hacky code. A really small win I know, but still something that might have been easily missed  when onboarding if kept within the main styles.

Usually, if I really feel something needs to be in a shame file, I will comment in the component stylesheet as well to say that (although, I admit I write a lot of comments and not everyone will do that). Mostly though I like to make sure my shame files are empty because then I know I haven't written anything I consider to be hacky.

Anyway, I know this is moving off topic so maybe this is a discussion for the Frontend guild instead :)

**Comment by peterwilsoncc** [comment]
@mikeselander @mattheu 

I don't think there's anything in the standards that encourage `!important`, rather that the type of code the standards encourage include legitimate uses. Those that come to mind are:

* utility classes
  ```css
  has-blue-color { color: #00f !important }
  ```
* user options
  ```css
  @media (prefers-reduced-motion: reduce) {
  * {
      transition: none !important
    }
  }
  ```
* everything in the WP default `.screen-reader-text`
* Some user facing options in the WP editor

I think it's just enough legit uses for such a rule to become a hindrance rather than a benefit.

**Comment by sambulance** [comment]
I also think we should be highly discouraging the use of `!important`. Most legitimate use cases are few and far between, so to me managing these is feasible. 

The editor is a bit of a trickier one, because we are already fighting the specificity of the WP editor styles - same with any third-party code we may need to adapt (Gravity Forms sucks for this too). You only really have two options - `!important`, or awfully specific selectors. Which is the lesser of the two evils? 

We have rules against specificity, but not `!important`. Personally, I think we should disallow both with the option to ignore the rule, as we currently have for specificity, and require a comment to explain why.

**Comment by peterwilsoncc** [comment]
> We have rules against specificity, but not !important. Personally, I think we should disallow both with the option to ignore the rule, as we currently have for specificity, and require a comment to explain why.

Convinced.
